# SQL Project &ndash; Book Reviews App

**Problem Statement:** The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.

This SQL project aims to study one such app towards creating a new value proposition for the app in terms of marketing positioning and brand salience.

# Contents <a id='back' ><a/>
    
* [1. Initialization](#initialization)  
* [2. Exploratory Data Analysis](#eda)      
    * [2.1. Books Released proceeding January 1st, 2000](#eda1)
    * [2.2. User Review Count and Average Rating by Unique Books](#eda2) 
    * [2.3. Book Releases by Publisher - Selecting the Author with most book releases](#eda3) 
    * [2.4. Average Rating by Author ](#eda4) 
    * [2.5. Average User Review Count](#eda5) 
* [3. Conclusions](#conclusions) 

# 1. Initialization <a id='initialization' ><a/>

In [11]:
!pip install pandas sqlalchemy
!pip install psycopg2-binary

In [12]:
# importing libraries

import pandas as pd
from sqlalchemy import create_engine

In [13]:
# Connecting to the database

db_config = {'user': 'praktikum_student',         # username
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [14]:
# Defining Functions

def execute_query_to_df(query, engine):
    with engine.connect() as conn:
        df = pd.io.sql.read_sql(query, conn)
    return df

In [15]:
# Reading tables [books, authors, ratings, reviews, publishers]

query_books = ''' SELECT * FROM books '''
query_authors = ''' SELECT * FROM authors '''
query_ratings = ''' SELECT * FROM ratings '''
query_reviews = ''' SELECT * FROM reviews '''  
query_publishers = ''' SELECT * FROM publishers '''


df_books = execute_query_to_df(query_books, engine)
df_authors = execute_query_to_df(query_authors, engine)
df_ratings = execute_query_to_df(query_ratings, engine)
df_reviews = execute_query_to_df(query_reviews, engine)
df_publishers = execute_query_to_df(query_publishers, engine)

In [16]:
display(df_books.head())
display(df_books.info())
display(df_authors.head())
display(df_authors.info())
display(df_ratings.head())
display(df_ratings.info())
display(df_reviews.head())
display(df_reviews.info())
display(df_publishers.head())
df_publishers.info()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


[Back to Contents](#back)

# 2. Exploratory Data Analysis <a id='eda' ><a/>

### 2.1. Books Released proceeding January 1st, 2000 <a id='eda1' ><a/>

In [17]:
# Querying book ID's released on or after the 1st of Jan, 2000

books_y2k_query = '''
    SELECT 
        COUNT(book_id)
    FROM 
        books
    WHERE
        publication_date > '2000-01-01';
'''

In [18]:
books_y2k = execute_query_to_df(books_y2k_query, engine)
print("Books after year 2000_Percentage Share:", books_y2k.loc[0, 'count'] / len(df_books))
books_y2k.head()

Books after year 2000_Percentage Share: 0.819


count
0    819

> **Observations:** Roughly 81% of books in the database were released post-year 2000.

[Back to Contents](#back)

### 2.2. User Review Count and Average Rating by Unique Books <a id='eda2' ><a/>

In [19]:
# Querying total reviews and average ratings by unique book ID's

ratings_reviews_query = '''
    SELECT
        reviews.book_id,
        books.title,
        COUNT(DISTINCT reviews.review_id) AS n_reviews,
        AVG(ratings.rating) AS avg_rating
    FROM 
        reviews 
    JOIN 
        ratings 
            ON reviews.book_id = ratings.book_id
    JOIN
        books
            ON reviews.book_id = books.book_id
    GROUP BY
        reviews.book_id, books.title
    ORDER BY
        n_reviews DESC, avg_rating DESC;
'''

In [20]:
ratings_reviews_by_book = execute_query_to_df(ratings_reviews_query, engine)
ratings_reviews_by_book.head(20)

book_id                                              title  n_reviews  \
0       948                            Twilight (Twilight  #1)          7   
1       302  Harry Potter and the Prisoner of Azkaban (Harr...          6   
2       299  Harry Potter and the Chamber of Secrets (Harry...          6   
3       656                                     The Book Thief          6   
4       734                                   The Glass Castle          6   
5       750                The Hobbit  or There and Back Again          6   
6       497                          Outlander (Outlander  #1)          6   
7       695  The Curious Incident of the Dog in the Night-Time          6   
8       779  The Lightning Thief (Percy Jackson and the Oly...          6   
9       963                                Water for Elephants          6   
10      696             The Da Vinci Code (Robert Langdon  #2)          6   
11      673                             The Catcher in the Rye          6   
12      627                                      The Alchemist          6   
13      854                                           The Road          6   
14      733                          The Giver (The Giver  #1)          6   
15      207                                    Eat  Pray  Love          6   
16      581                     Snow Flower and the Secret Fan          5   
17       74                           And Then There Were None          5   
18      884                                The Thirteenth Tale          5   
19      284                                 Gone with the Wind          5   

    avg_rating  
0     3.662500  
1     4.414634  
2     4.287500  
3     4.264151  
4     4.206897  
5     4.125000  
6     4.125000  
7     4.081081  
8     4.080645  
9     3.977273  
10    3.830508  
11    3.825581  
12    3.789474  
13    3.772727  
14    3.750000  
15    3.395833  
16    4.636364  
17    4.454545  
18    4.444444  
19    4.428571

> **Observations:** We can observe a trend of disproportionate correlation between the number of reviews and average ratings of a book. We can clearly see the sheer popularity of the first Twilight book judging from the number of reviews written for it, but it did not necessarily equate to anywhere near the best average rating. Hence, the number of reviews are a better indicator of a book's popularity; this could be due to the influence of pop culture, as seen in the presence of 4 different *Harry Potter* sequels and two *Lord of The Rings* books (The Hobbit included), while we also see classics such as *Romeo and Juliet* and Internation Best-Sellers in *The Book Thief* in our top 20 list. All in all, a great blend of classics, pop-culture and modern-day best-sellers.

[Back to Contents](#back)

### 2.3. Book Releases by Publisher - Selecting the Author with most book releases <a id='eda3' ><a/>

**N.B.** Filtering for books above 50 pages to eliminate occurrences of brochures and similar-natured publications.

In [21]:
# Querying publisher Id's and their corresponding book counts, filtered for no. of pgs above 50

publisher_books_query = '''
    SELECT
        publishers.publisher, publishers.publisher_id,
        COUNT(books.book_id) as book_releases
    FROM 
        books 
    JOIN publishers 
        ON books.publisher_id = publishers.publisher_id
    WHERE
        books.num_pages > 50
    GROUP BY
        publishers.publisher, publishers.publisher_id
    ORDER BY
        book_releases DESC
    LIMIT 1;
'''

In [22]:
releases_by_publisher = execute_query_to_df(publisher_books_query, engine)
releases_by_publisher.sort_values('book_releases', ascending=False).head(10)

publisher  publisher_id  book_releases
0  Penguin Books           212             42

> **Observations:** The Penguin brand is by and wide the largest publisher on display here, especially when we consider its main affordable brand, Penguin Books and its more premium imprint sub-brand, Penguin Classics made it into our top 10. This is no surprise however, as Penguin Publications are renowned for being the leading publisher of classic literature in the English-speaking world. All time classics such as Price & Prejudice, Moby Dick and Dracula were all published via the Penguin brand.

[Back to Contents](#back)

### 2.4. Average Rating by Author <a id='eda4' ><a/>

**N.B.** Filtering for books with at least 50 n_ratings.

In [195]:
# Querying average rating by author, for books with at least 50 ratings

average_rating_author_query = '''
    SELECT
        authors.author,
        AVG(ratings.rating) AS avg_rating
    FROM
        authors
        JOIN books ON authors.author_id = books.author_id
        JOIN ratings ON books.book_id = ratings.book_id
    WHERE
        books.book_id IN (
            SELECT
                book_id
            FROM
                ratings
            GROUP BY
                book_id
            HAVING
                COUNT(rating_id) >= 50
        )
    GROUP BY
        authors.author
    ORDER BY
        avg_rating DESC;
'''

In [196]:
authors_by_rating = execute_query_to_df(average_rating_author_query, engine)
authors_by_rating.head(5)

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.287097
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.246914
3                  Louisa May Alcott    4.192308
4                       Rick Riordan    4.080645

> **Observations:** To any book-fan, the top 3 authors by average ratings is no surprise. The author J.K. Rowling (Mary GrandPre' being the illustrator) is best known for her series of *Harry Potter* books including other sub-series of the Harry Potter universe. Markus Zusak on the other hand wrote *The Book Thief* and *The Messenger*, two books that went on to become Internation Best-Sellers. Lastly, who could forget J.R.R. Tolkien with *Lord of The Rings* franchise. This also includes works  in and of the same sphere of the likes of the Hobbit and Silmarillion.

[Back to Contents](#back)

### 2.5. Average User Review Count <a id='eda5' ><a/>

**N.B.** To aggregate review counts for only users who rated books on a high-frequency basis, we will filter for n_ratings above 50. We'd like to observe the behaviour of an avid-book reader and reviewer.

In [171]:
agg_review_count_query = '''
    SELECT AVG(review_count) AS avg_text_reviews
        FROM (
            SELECT 
                username, COUNT(review_id) AS review_count
            FROM (
                SELECT 
                    reviews.username, reviews.review_id
                FROM 
                    reviews 
                JOIN 
                    ratings ON reviews.username = ratings.username
                GROUP BY 
                    reviews.username, reviews.review_id
                HAVING 
                    COUNT(DISTINCT ratings.book_id) > 50
                    ) AS subquery1
            GROUP BY username) AS subquery2;
'''

In [172]:
avg_reviews_by_user = execute_query_to_df(agg_review_count_query, engine)
avg_reviews_by_user.head(5)

avg_text_reviews
0         24.333333

> **Observations:** An avid reader and subsequently, reviewer, who rates books above the 50 book threshold typically writes roughly 24 reviews in their lifetime (and counting). These readers can be considered major book-worms, and are seen as crucial towards word-of-mouth marketing on a book-by-book basis.

[Back to Contents](#back)

# 3. Conclusions <a id='conclusions' ><a/>
    
**Value Proposition moving forward as a Books Review site:**   
    
1. 82% of our books on observation are post-year 2000, which in contrast, means that we still have a fair share of books that are deemed 'older classics' on display. This could be major boon to avid readers of an older nature and literature buffs alike.
2. Our top 10 snippet of books with high review counts and moderate-to-high ratings tells us that we have a wide array of popular books on display, which ranges from all-time classics, popular culture influenced titles and modern-day Internation Best-Sellers. This bodes well for readers from all walks of life and thus, could be a one-stop centre for anyone vying to pick up their next book.
3. It is encouraging to see the large majority of books on display being from the Penguins publishing brand, as they are by far, the largest literature-based publisher in the English-speaking world. This is a great selling point, as the app prides itself in catering to books that are, for lack of a better phrase, *'good reads'*.
4. Our top authors on the app, with an average rating of above 4, are inherently popular with the masses. These being authors such as J.K. Rowling, J.R.R. Tolkien and Markus Zusak. The is important as more likely than not, titles from these authors would inevitably be recommended sooner or later down the line. Furthermore, this creates a domino effect for lesser known titles.
5. Our app in question contains a subset of unique users who help generate word-of-mouth traffic, in part because they frequently rate books at a high clip (above 50 distinct books), while also write reviews on a frequent basis (average of roughly 24 reviews). These users help create relevancy to the app, as well as help promote back-link generation for book sales. This is a huge value proposition, as it means the app has potential (if it hasn't already reached it) towards being a one-stop centre any information book-related, or rather, an intermediary connecting all types of readers to the masses of publishers and sellers. 